android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/Utils.java

```diff
@@ -253,23 +253,46 @@ private boolean needApiKey() {
 
   }
 
-  private static class ClearCacheDiscoveryServiceConnection implements ServiceConnection {
+  private static class DiscoveryServiceConnection implements ServiceConnection {
+    private static final int ACTION_CLEAR_CACHE = 0;
+    private static final int ACTION_START_SCAN = 1;
     private Context mContext;
+    private int mRequestedAction;
 
     @Override
     public synchronized void onServiceConnected(ComponentName className, IBinder service) {
       // Get the service
       UrlDeviceDiscoveryService.LocalBinder localBinder =
           (UrlDeviceDiscoveryService.LocalBinder) service;
-      localBinder.getServiceInstance().clearCache();
+      UrlDeviceDiscoveryService scanner = localBinder.getServiceInstance();
+      switch (mRequestedAction) {
+        case ACTION_CLEAR_CACHE:
+          scanner.clearCache();
+          break;
+        case ACTION_START_SCAN:
+          scanner.newPwsStartScan();
+          break;
+        default:
+          break;
+      }
       mContext.unbindService(this);
     }
 
     @Override
     public synchronized void onServiceDisconnected(ComponentName className) {
     }
 
-    public synchronized void connect(Context context) {
+    public synchronized void clearCache(Context context) {
+      mRequestedAction = ACTION_CLEAR_CACHE;
+      connect(context);
+    }
+
+    public synchronized void startScan(Context context) {
```

Comment:
Why not just created a new kind of connection in addition to the ClearCacheDiscoveryServiceConnection?  Then you wouldn't need to do any synchronization.